File: typescript/blueprints/url-shortener/lib/url-shortener-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export class UrlShortenerStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // DynamoDB table for storing URL mappings
    const urlTable = new dynamodb.Table(this, 'UrlTable', {
      tableName: 'url-shortener',
      partitionKey: { name: 'shortCode', type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // Add GSI for original URL lookups
    urlTable.addGlobalSecondaryIndex({
      indexName: 'originalUrl-index',
      partitionKey: { name: 'originalUrl', type: dynamodb.AttributeType.STRING },
    });

    // Lambda function for URL shortening
    const shortenFunction = new lambda.Function(this, 'ShortenFunction', {
      runtime: lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('lambda/shorten'),
      environment: {
        TABLE_NAME: urlTable.tableName,
      },
    });

    // Lambda function for URL redirection
    const redirectFunction = new lambda.Function(this, 'RedirectFunction', {
      runtime: lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('lambda/redirect'),
      environment: {
        TABLE_NAME: urlTable.tableName,
      },
    });

    // Lambda function for analytics
    const analyticsFunction = new lambda.Function(this, 'AnalyticsFunction', {
      runtime: lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset('lambda/analytics'),
      environment: {
        TABLE_NAME: urlTable.tableName,
      },
    });

    // Grant DynamoDB permissions
    urlTable.grantReadWriteData(shortenFunction);
    urlTable.grantReadWriteData(redirectFunction);
    urlTable.grantReadData(analyticsFunction);

    // API Gateway
    const api = new apigateway.RestApi(this, 'UrlShortenerApi', {
      restApiName: 'URL Shortener Service',
      description: 'API for URL shortening service',
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowMethods: apigateway.Cors.ALL_METHODS,
        allowHeaders: ['Content-Type', 'X-Amz-Date', 'Authorization', 'X-Api-Key'],
      },
    });

    // API routes
    const shortenIntegration = new apigateway.LambdaIntegration(shortenFunction);
    const redirectIntegration = new apigateway.LambdaIntegration(redirectFunction);
    const analyticsIntegration = new apigateway.LambdaIntegration(analyticsFunction);

    // POST /shorten - Create short URL
    api.root.addResource('shorten').addMethod('POST', shortenIntegration);

    // GET /{shortCode} - Redirect to original URL
    const shortCodeResource = api.root.addResource('{shortCode}');
    shortCodeResource.addMethod('GET', redirectIntegration);

    // GET /analytics/{shortCode} - Get analytics for short URL
    const analyticsResource = api.root.addResource('analytics').addResource('{shortCode}');
    analyticsResource.addMethod('GET', analyticsIntegration);

    // S3 bucket for static website
    const websiteBucket = new s3.Bucket(this, 'WebsiteBucket', {
      websiteIndexDocument: 'index.html',
      websiteErrorDocument: 'error.html',
      publicReadAccess: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // CloudFront distribution
    const distribution = new cloudfront.Distribution(this, 'Distribution', {
      defaultBehavior: {
        origin: new origins.S3Origin(websiteBucket),
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      additionalBehaviors: {
        '/api/*': {
          origin: new origins.RestApiOrigin(api),
          viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
          cachePolicy: cloudfront.CachePolicy.CACHING_DISABLED,
        },
      },
    });

    // Outputs
    new cdk.CfnOutput(this, 'ApiUrl', {
      value: api.url,
      description: 'API Gateway URL',
    });

    new cdk.CfnOutput(this, 'DistributionUrl', {
      value: `https://${distribution.distributionDomainName}`,
      description: 'CloudFront Distribution URL',
    });

    new cdk.CfnOutput(this, 'WebsiteBucketName', {
      value: websiteBucket.bucketName,
      description: 'S3 Website Bucket Name',
    });
  }
}

File: python/blueprints/url_shortener/url_shortener_stack.py
import aws_cdk as cdk
from aws_cdk import (
    Stack,
    aws_lambda as _lambda,
    aws_dynamodb as dynamodb,
    aws_apigateway as apigateway,
    aws_cloudfront as cloudfront,
    aws_cloudfront_origins as origins,
    aws_s3 as s3,
)
from constructs import Construct


class UrlShortenerStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # DynamoDB table for storing URL mappings
        url_table = dynamodb.Table(
            self, "UrlTable",
            table_name="url-shortener",
            partition_key=dynamodb.Attribute(
                name="shortCode", 
                type=dynamodb.AttributeType.STRING
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            removal_policy=cdk.RemovalPolicy.DESTROY,
        )

        # Add GSI for original URL lookups
        url_table.add_global_secondary_index(
            index_name="originalUrl-index",
            partition_key=dynamodb.Attribute(
                name="originalUrl", 
                type=dynamodb.AttributeType.STRING
            ),
        )

        # Lambda function for URL shortening
        shorten_function = _lambda.Function(
            self, "ShortenFunction",
            runtime=_lambda.Runtime.PYTHON_3_9,
            handler="lambda_function.handler",
            code=_lambda.Code.from_asset("lambda/shorten"),
            environment={
                'TABLE_NAME': url_table.table_name,
            },
        )

        # Lambda function for URL redirection
        redirect_function = _lambda.Function(
            self, "RedirectFunction",
            runtime=_lambda.Runtime.PYTHON_3_9,
            handler="lambda_function.handler",
            code=_lambda.Code.from_asset("lambda/redirect"),
            environment={
                'TABLE_NAME': url_table.table_name,
            },
        )

        # Lambda function for analytics
        analytics_function = _lambda.Function(
            self, "AnalyticsFunction",
            runtime=_lambda.Runtime.PYTHON_3_9,
            handler="lambda_function.handler",
            code=_lambda.Code.from_asset("lambda/analytics"),
            environment={
                'TABLE_NAME': url_table.table_name,
            },
        )

        # Grant DynamoDB permissions
        url_table.grant_read_write_data(shorten_function)
        url_table.grant_read_write_data(redirect_function)
        url_table.grant_read_data(analytics_function)

        # API Gateway
        api = apigateway.RestApi(
            self, "UrlShortenerApi",
            rest_api_name="URL Shortener Service",
            description="API for URL shortening service",
            default_cors_preflight_options=apigateway.CorsOptions(
                allow_origins=apigateway.Cors.ALL_ORIGINS,
                allow_methods=apigateway.Cors.ALL_METHODS,
                allow_headers=["Content-Type", "X-Amz-Date", "Authorization", "X-Api-Key"],
            ),
        )

        # API routes
        shorten_integration = apigateway.LambdaIntegration(shorten_function)
        redirect_integration = apigateway.LambdaIntegration(redirect_function)
        analytics_integration = apigateway.LambdaIntegration(analytics_function)

        # POST /shorten - Create short URL
        api.root.add_resource("shorten").add_method("POST", shorten_integration)

        # GET /{shortCode} - Redirect to original URL
        short_code_resource = api.root.add_resource("{shortCode}")
        short_code_resource.add_method("GET", redirect_integration)

        # GET /analytics/{shortCode} - Get analytics for short URL
        analytics_resource = api.root.add_resource("analytics").add_resource("{shortCode}")
        analytics_resource.add_method("GET", analytics_integration)

        # S3 bucket for static website
        website_bucket = s3.Bucket(
            self, "WebsiteBucket",
            website_index_document="index.html",
            website_error_document="error.html",
            public_read_access=True,
            block_public_access=s3.BlockPublicAccess(
                block_public_acls=False,
                block_public_policy=False,
                ignore_public_acls=False,
                restrict_public_buckets=False,
            ),
            removal_policy=cdk.RemovalPolicy.DESTROY,
        )

        # CloudFront distribution
        distribution = cloudfront.Distribution(
            self, "Distribution",
            default_behavior=cloudfront.BehaviorOptions(
                origin=origins.HttpOrigin(website_bucket.bucket_website_domain_name),
                viewer_protocol_policy=cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
            ),
            additional_behaviors={
                "/api/*": cloudfront.BehaviorOptions(
                    origin=origins.RestApiOrigin(api),
                    viewer_protocol_policy=cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
                    cache_policy=cloudfront.CachePolicy.CACHING_DISABLED,
                ),
            },
        )

        # Outputs
        cdk.CfnOutput(
            self, "ApiUrl",
            value=api.url,
            description="API Gateway URL",
        )

        cdk.CfnOutput(
            self, "DistributionUrl",
            value=f"https://{distribution.distribution_domain_name}",
            description="CloudFront Distribution URL",
        )

        cdk.CfnOutput(
            self, "WebsiteBucketName",
            value=website_bucket.bucket_name,
            description="S3 Website Bucket Name",
        )
